### ISSUE TITLE
How to prevent operator from fusing when converting to TFLite model? 

### ISSUE THREAD

**Issue by yxchng** [tensorflow/tensorflow#35176]
GPU delegate does not support operator fusion, so I need to prevent the conversion from fusing the operators.

TFLiteConvertor will automatically fuse operators, which gives problems.

**Comment by karimnosseir** [comment]
Hi,

Currently the only way is to change the code to skip doing this.
For example,
if you want to stop fusing Activation function
- In TOCO
  It is defined here
 https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/toco/graph_transformations/fuse_activation_functions.cc
 to disable specific op from fusing  you can change
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/toco/tooling_util.cc#L469

- In New Converter
You need to find the pattern that does this transform.
For binary ops the pattern is here
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/mlir/lite/transforms/optimize_patterns.td#L297

You can remove the type you don't want or remove the whole pattern completely.

Hope that helps.

Thanks

**Comment by yxchng** [comment]
@karimnosseir Is there any way to not touch MLIR code? I can't really understand MLIR code.

**Comment by karimnosseir** [comment]
Hi @yxchng The current default behavior uses TOCO, so you don't need to touch MLIR code (yet).

It will be good to hear from you and others as well, how can we make our new MLIR converter easy for the community to understand and contribute.
Did you check the guide https://github.com/tensorflow/mlir#getting-started-with-mlir ?

Thanks

**Comment by yxchng** [comment]
@karimnosseir I have checked them but it is still too complicated for me. I have no idea of compilers and LLVM. It is not written in a way that is accessible to people with 0 knowledge of compilers.

**Comment by ymodak** [comment]
Closing this issue since its resolved. Feel free to reopen if have further questions. Thanks!

**Comment by dimitraka** [comment]
I would like to disable fusion from TensorFlow's code, so is there any easy way to do so?

I've already checked this file 
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/mlir/lite/transforms/optimize_patterns.td
but since I'm not familiar with LLVM, i's kind of hard for me.

Could you maybe give an example of how to prevent fusion for Conv2D and ReLU let's say? Which lines of code should I remove in this case?

Thanks a lot!

**Comment by karimnosseir** [comment]
Hi @dimitraka, 
Thanks for the feedback, we will look in making this easier for users to customize.
For now this is what triggers the fusing of Conv2D with relu you can comment them and rebuild
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/mlir/lite/transforms/optimize_patterns.td#L84:L89

**Comment by dimitraka** [comment]
Thank you for the immediate answer!

Could you tell me how to rebuild as well? Because I have some trouble finding out how to rebuild TensorFlow once I have installed it from source?

I tried the following:
`bazel build -j 16 //tensorflow/compiler/mlir/lite:tensorflow_lite_optimize` 
for building only a specific target.

I also tried building TensorFlow, like when first installing it.
`
bazel build -j 16 //tensorflow/tools/pip_package:build_pip_package
`
`
./bazel-bin/tensorflow/tools/pip_package/build_pip_package /tmp/tensorflow_pkg
`
`
pip install /tmp/tensorflow_pkg/tensorflow-2.3.0-cp38-cp38m-linux_x86_64.whl
`

But didn't see any difference in my graph in any way. Do you have any suggestions?

**Comment by karimnosseir** [comment]
Hi @dimitraka, 
You can build the converter command line and run it.

bazel build -c opt tensorflow/lite/python:tflite_convert

then run it
bazel-bin/tensorflow/lite/python/tflite_convert --saved_model_dir=<path> --output_file=/tmp/my_model.tflite

**Comment by manp-git** [comment]
hi @karimnosseir 
I am only aware of building tensorflow lite using :-
tensorflow.lite.TFLiteConverter.from_keras_model(my_model)
and the apply converter optimizations.

Will this method also do the same job after commenting the mentioned lines in 
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/mlir/lite/transforms/optimize_patterns.td#L84:L89

Hi @dimitraka is your issue resolved?

**Comment by manp-git** [comment]
Also I am unable to find the batch normalisation layer in the TFLite model that is supposed to be in between convolution and activation layer in a Resnet50 model, what happened to that? It fused between convolution and Relu?

**Comment by FraCorti** [comment]
Hi, are there any news regarding this?
